### ISSUE TITLE
Standardise module

### ISSUE THREAD

**Issue by emteknetnz** [silverstripe/silverstripe-linkfield#78]
Rough list of things we could do to this module to get it in line with other Silverstripe modules

Note will release an new major so can make API breaks:

Note there's a second look details arou

# Task list

## Architecture
- See https://github.com/silverstripe/silverstripe-linkfield/issues/81

## Code quality
- [x] Add strong typing
- [x] Look at unit tests coverage and add more if it makes sense to
- [x] Consider turning the two LinkField-related traits into an abstract class instead, which `LinkField` and `MultiLinkField` can be subclasses of
  - [x] Look at other code duplication that can also be moved into that abstract class
- [x] Delete unused code (e.g. `client/src/LinkModal/FileLinkModal.js`)

## Maintenance
- [x] Rename .github/workflows/main.yml to .github/workflows/ci.yml
- [x] Run module standardiser, will add things like dispatch-ci.yml and keepalive.yml
- [x] Update module standardiser to remove other old meta files e.g. .codeclimate.yml etc
- [x] Update readme badges, get them consistent with other modules
- [x] Fix typos e.g. LinkTypeResolver.php "provdied"
- [x] Follow instruction for [creating new repo](https://silverstripe.atlassian.net/wiki/spaces/CMS/pages/3030679981/Setting+up+a+new+GitHub+repository).
- [x] Add to transifex
- [x] Add to kitchen sink

## PRs
- https://github.com/silverstripe/silverstripe-linkfield/pull/100
- https://github.com/silverstripe/silverstripe-admin/pull/1678
- https://github.com/silverstripe/silverstripe-linkfield/pull/216
- https://github.com/silverstripe/module-standardiser/pull/38
- https://github.com/silverstripe/supported-modules/pull/27
- https://github.com/silverstripe/recipe-kitchen-sink/pull/48
- https://github.com/silverstripe/silverstripe-linkfield/pull/223

**Comment by maxime-rainville** [comment]
Just a bit of extra context.

## DBLink
The DBLink implementation was something I put in there because @chillu was concerned adding tables for each link type would clog the database. I for one never was a big fan of this.

I think the fact that most people naturally went for the Link DataObject shows that most dev prefer this approach because it's closer to what they are used to.

If we ditch DBLink there's a lot of extra simplication we can put in place. The following classes can be ditch altogether:
- [LinkTypeResolver](https://github.com/silverstripe/silverstripe-linkfield/blob/3/src/GraphQL/LinkTypeResolver.php)
- [DBJson](https://github.com/silverstripe/silverstripe-linkfield/blob/3/src/ORM/DBJson.php)
- [DBLink](https://github.com/silverstripe/silverstripe-linkfield/blob/3/src/ORM/DBLink.php)
- [Registry](https://github.com/silverstripe/silverstripe-linkfield/blob/3/src/Type/Registry.php)
- [Type](https://github.com/silverstripe/silverstripe-linkfield/blob/3/src/Type/Type.php)
- [JsonData](https://github.com/silverstripe/silverstripe-linkfield/blob/3/src/JsonData.php)


## Relationship to Any Field

With the [AnyField module](https://github.com/maxime-rainville/silverstripe-anyfield), I took the LinkField logic and made it more generic so it could be used to manage any DataObject.

I think there's a strong case to make that AnyField is better in every possible way than the current linkfield.
- I can manage list of links as well as single links.
- It has better UI (support for icons and loading animation).
- It allows you to constrain what type of links can be created in a specific instance.
- Once this [PR is merged](https://github.com/maxime-rainville/silverstripe-anyfield/pull/22), it will allow you to reorder things.

I've also added pretty comprehensive PHPUnit and Behat test to the AnyField using Links as a test model.

I'm thinking, either:
- we just tell people to use the AnyField to manage links or,
- we implement a LinkField that simply extends the regular AnyField with a bit of extra Link specific functionality.

## Migration
Given the little adoption of DBLink, I don't think there's much value in providing a migration task.

There could be a better case to provide a migration task from other module to this one.

## Other concerns
- Permission for Link haven't really quite been thought through.
- Not sure how we should handle versioning.
- I've got a [bunch of related cards on the AnyField module](https://github.com/maxime-rainville/silverstripe-anyfield/issues).
- https://github.com/silverstripe/silverstripe-linkfield/pull/76 adds a few bells and whistle to make this module work well with AnyField.

**Comment by emteknetnz** [comment]
I dunno about Anyfield given the context here is linkfield support.  I don't doubt that the core idea of anyfield could be useful (as I understand it the core thing is that you can manage DataObject's in a modal rather than on a regular form, correct me if I'm wrong).

It's just the context here is we're looking at bumping up the support given to the linkfield module which already has a pretty reasonable amount of real world usage.

Seems that bringing in anyfield is increasing the scope to "better support managing dataobjects in a modal.".  Wouldn't that sort of functionality should maybe be added directly in silverstripe/admin rather than a module called silverstripe/linkfield?

**Comment by emteknetnz** [comment]
Some of the things on this issue where implemented here https://github.com/silverstripe/silverstripe-linkfield/issues/84

**Comment by maxime-rainville** [comment]
I'll put this card up for refinement, but it seems way too big to be done in one go. So we'll probably split it up.

**Comment by GuySartorelli** [comment]
Reopening because I found some more things that aren't quite standardised that I somehow missed last time.
